Bluetooth: Standardize radio HW capability representation in devicetree

ipankr opened the pull request:
Summary
-------
This PR creates a shared base binding for common Bluetooth LE radio hardware capabilities to avoid duplication between vendors and ensure consistent property naming across the Zephyr ecosystem.

Changes
-------

1. New Base Binding
   - Created dts/bindings/bluetooth/ble-radio.yaml with common BLE hardware properties
   - Properties prefixed with ble- for consistency
   - Ordered chronologically by Bluetooth Core Specification version (5.0, 5.1, 6.0)

2. Updated Vendor Bindings
   
   Silabs:
   - Updated silabs,series2-radio.yaml to include base binding
   - Kept vendor-specific ble-tx-high-power-supported property
   
   Nordic:
   - Updated nordic,nrf-radio.yaml to include base binding
   - Renamed tx-high-power-supported to ble-tx-high-power-supported for consistency

3. Silicon Labs Device Trees
   
   Updated all Series 2 SoCs with hardware-supported features:
   - EFR32xG21: 2M PHY, Coded PHY, TX High Power
   - EFR32xG22: 2M PHY, Coded PHY, CTE TX, CTE RX, TX High Power
   - EFR32xG24: 2M PHY, Coded PHY, CTE TX, CTE RX, Channel Sounding, TX High Power
   - EFR32xG27: 2M PHY, Coded PHY, CTE TX, TX High Power
   - EFR32xG29: 2M PHY, Coded PHY, CTE TX, TX High Power

4. Nordic Device Trees
   
   Updated with consistent property naming:
   - nRF52820, nRF52833, nRF52840: Added ble-tx-high-power-supported
   - nRF54H20, nRF54L15, nRF54LM20: Renamed to ble-cs-supported

5. Kconfig Integration
   
   Silabs - Added hardware capability flags (auto-detected from DT):
   - HAS_HW_EFR32_RADIO_BLE_2M
   - HAS_HW_EFR32_RADIO_BLE_CODED
   - HAS_HW_EFR32_RADIO_CTE_TX
   - HAS_HW_EFR32_RADIO_CTE_RX
   - HAS_HW_EFR32_RADIO_CS
   - HAS_HW_EFR32_RADIO_TX_HIGH_POWER
   
   Silabs HCI driver automatically selects controller features based on hardware capabilities.
   
   Nordic - Updated Kconfig to reference renamed properties:
   - HAS_HW_NRF_RADIO_CS uses ble-cs-supported
   - HAS_HW_NRF_RADIO_TX_PWR_HIGH uses ble-tx-high-power-supported

Testing
-------
Verified on all affected hardware (11 board configurations):
- All Silabs: XG21, XG22, XG24, XG27, XG29
- All Nordic: nRF52820, nRF52833, nRF52840, nRF54H20, nRF54L15, nRF54LM20

Design Philosophy
-----------------
This approach harmonizes devicetree bindings across vendors while respecting each vendor's driver architecture.

ipankr committed:
drivers: bluetooth: efr32: move Silabs Kconfig settings

Vendor specific configuration is moved to a separate file.

Signed-off-by: Ivan Pankratov <ivan.pankratov@silabs.com>

ipankr committed:
dts: bluetooth: add base binding for BLE radio hardware capabilities

Create a shared base binding (ble-radio.yaml) for common Bluetooth LE
radio hardware capabilities to avoid duplication between vendors and
ensure consistent property naming across the ecosystem.

Properties are prefixed with 'ble-' and ordered chronologically by
Bluetooth Core Specification version (5.0, 5.1, 6.0). Each property
indicates a hardware capability, not current enablement state.

Signed-off-by: Ivan Pankratov <ivan.pankratov@silabs.com>

jhedberg commented on dts/bindings/net/wireless/silabs,series2-radio.yaml:
```diff
@@ -56,3 +56,43 @@ properties:
+    description: |
+      If set, indicates that the radio hardware supports coded BLE PHY.
+      This property should be treated as read-only and should not be overridden;
+      the correct value is provided for your target's SoC already.
```
Since we're close to 4.3, I don't think I'd like to have a mix there where one radio binding uses the new base binding and another doesn't, so I'd suggest to update the nrf-radio.yaml here as well, if you don't mind (for what I've seen, it's just two ble- properties, right?)

jhedberg commented on dts/bindings/net/wireless/silabs,series2-radio.yaml:
```diff
@@ -56,3 +56,12 @@ properties:
     enum:
       - highest
+
+  ble-tx-high-power-supported:
```
Even though this isn't based on any spec, but since there's a property with the identical name in the nRF binding, perhaps we'd still want this in the common base binding to avoid duplication (with a mention in the description that the meaning is implementation-specific?). Thoughts from @cvinayak?

cvinayak replied:
see https://github.com/zephyrproject-rtos/zephyr/pull/97994/files#r2451718718

`tx-high-power-supported` is very generic to wireless.  Can we or do we already have a `generic,radio.yaml` ?  this should go there.

cvinayak commented on dts/vendor/nordic/nrf54h20.dtsi:
```diff
@@ -488,7 +488,7 @@
 				ble-coded-phy-supported;
-				cs-supported;
+				ble-cs-supported;
 				dfe-supported;
```
`dfe-supported`  ? `ble-dfe-supported`, right?

cvinayak replied:
@ipankr could you address this as well, thanks.

nordicjm commented on drivers/bluetooth/hci/Kconfig.silabs:
```diff
@@ -1,6 +1,60 @@
+	select BT_CTLR_DF_ANT_SWITCH_1US_SUPPORT if HAS_HW_EFR32_RADIO_CTE_TX
+	select BT_CTLR_DF_ANT_SWITCH_2US_SUPPORT if HAS_HW_EFR32_RADIO_CTE_TX
+	select BT_SILABS_EFR32_HIGH_POWER if HAS_HW_EFR32_RADIO_TX_HIGH_POWER
+
```
no thanks, remove, not allowed under https://docs.zephyrproject.org/latest/contribute/style/kconfig.html

ipankr replied:
@nordicjm, addressed now

ipankr commented on soc/silabs/silabs_s2/Kconfig:
```diff
@@ -32,4 +32,23 @@ config SILABS_DEVICE_IS_MODULE
+	def_bool $(dt_nodelabel_bool_prop,radio,ble-2mbps-supported)
+
+config HAS_HW_EFR32_RADIO_BLE_CODED
+	def_bool $(dt_nodelabel_bool_prop,radio,ble-coded-phy-supported)
```
@jerome-pouiller please see [my comment](https://github.com/zephyrproject-rtos/zephyr/pull/97994#discussion_r2477921495)

jhedberg commented on dts/bindings/net/wireless/silabs,series2-radio.yaml:
```diff
@@ -56,3 +56,12 @@ properties:
+      This is a vendor-specific capability that enables transmission beyond
+      standard power levels (typically >10 dBm).
+      This property should be treated as read-only and should not be overridden;
+      the correct value is provided for your target's SoC already.
```
@cvinayak sounds good, but why `generic,` prefix? Based on the DT spec prefixes should be "identifying the name of the company or organizationthatdefined the property". Since Zephyr already has a convention of using the `zephyr,` prefix for non-vendor specific things, shouldn't we just go with that?

cvinayak replied:
> simply radio- for the radio binding would be fine?

Fine by me, if it is `radio-tx-high-power-supported` in `radio.yaml` , others Bluetooth specific continue to be `ble-2mbps-supported` in `ble-radio.yaml` , right? 

Additionally, rename other files `radio-fem-two-ctrl-pins.yaml`, `radio-gpio-coex.yaml`?

ipankr replied:
yes, updated according to the comments

ipankr committed:
dts: bindings: Rename generic-fem-two-ctrl-pins to radio-fem-two-ctrl-pins

Rename generic FEM binding to use 'radio-' prefix for consistency with
other radio bindings (radio.yaml, ble-radio.yaml).

Updated 3 board files, Nordic BLE controller HAL, documentation, and
test overlays. Added missing settle-time properties to pan1783a board.

Signed-off-by: Ivan Pankratov <ivan.pankratov@silabs.com>

ipankr committed:
dts: bindings: Rename gpio-radio-coex to radio-gpio-coex

Rename GPIO coex binding to use 'radio-' prefix for consistency with
other radio bindings (radio.yaml, radio-fem-two-ctrl-pins.yaml).

Updated beacon sample overlay and coex documentation.

Signed-off-by: Ivan Pankratov <ivan.pankratov@silabs.com>

ipankr committed:
doc: releases: document radio binding renames in 4.4

Add migration guide and release notes entries for the radio-related
devicetree binding renames:
- generic-fem-two-ctrl-pins -> radio-fem-two-ctrl-pins
- gpio-radio-coex -> radio-gpio-coex
- tx-high-power-supported -> radio-tx-high-power-supported

Also document the new radio.yaml base binding for generic radio
hardware capabilities.

Signed-off-by: Ivan Pankratov <ivan.pankratov@silabs.com>

ipankr commented on drivers/bluetooth/hci/Kconfig.silabs:
```diff
@@ -60,11 +114,11 @@ config BT_SILABS_EFR32_MAX_QUEUED_ADV_REPORTS
 	  Additional advertising reports are dropped.
 
 config BT_SILABS_EFR32_HIGH_POWER
```
Hi @nordicjm  I think I've addressed all your review comments. Could you please take another look.

jhedberg commented:
@nordicjm your concerns should (hopefully) finally be addressed. Could you please take one more look? Thanks.
